### PR TITLE
Bash

### PR DESCRIPTION
**Pull request by bootstraponline**
Fix https://github.com/TestArmada/flank/issues/200

### CODE REVIEW

File: cloud_testing/xctest_parser/src/main/kotlin/xctest/Bash.kt [hunk 1]
```diff
@@ -22,10 +31,19 @@ object Bash {
...
-        process.waitFor(2, TimeUnit.MINUTES)
+        val pool = Executors.newFixedThreadPool(2)
+        val gobbleInput = StreamGobbler(process.inputStream)
+        val gobbleError = StreamGobbler(process.errorStream)
+        pool.submit(gobbleInput)
+        pool.submit(gobbleError)
+
+        process.waitFor(1, TimeUnit.MINUTES)
 
         if (process.failed()) throw RuntimeException("Command failed: $cmd")
```
Comment (Macarse) [inline cloud_testing/xctest_parser/src/main/kotlin/xctest/Bash.kt]: I think we should do something like:
```
if (process.failed()) {
  println("error $gobbleError.output")
  throw RuntimeException("Command failed: $cmd")
}
```

File: cloud_testing/xctest_parser/src/main/kotlin/xctest/Bash.kt [hunk 2]
```diff
@@ -22,10 +31,19 @@ object Bash {
...
 
-        return process.output()
+        pool.shutdown()
+        pool.awaitTermination(1, TimeUnit.MINUTES)
+
+        return (gobbleInput.output + gobbleError.output).trim()
```
Comment (Macarse) [inline cloud_testing/xctest_parser/src/main/kotlin/xctest/Bash.kt]: if we get to this point, we should have only `gobbleInput.output`

File: cloud_testing/xctest_parser/src/main/kotlin/xctest/Bash.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+                .redirectOutput(PIPE)
+                .redirectError(PIPE)
+                .start()
+
+        val gobbleInput = StreamGobbler(process.inputStream)
+        gobbleInput.start()
+        val gobbleError = StreamGobbler(process.errorStream)
+        gobbleError.start()
+
+        process.waitFor()
```
Comment (Macarse) [inline cloud_testing/xctest_parser/src/main/kotlin/xctest/Bash.kt]:
```
        if (!process.waitFor(2, TimeUnit.MINUTES)) {
            process.destroy()
            throw RuntimeException("execution timed out: $this")
        }
```

**Comment by bootstraponline** [comment]
> why the timeout for the join()?

default join is 

```java
 public final void join() throws InterruptedException {
        join(0);
    }
```

which waits forever.

```
  * Waits at most {@code millis} milliseconds for this thread to
  * die. A timeout of {@code 0} means to wait forever.
```

**Comment by Macarse** [comment]
@bootstraponline ok, what I am saying is that we should be fine with `join()` since the process would be dead already.

**Comment by bootstraponline** [comment]
> @bootstraponline ok, what I am saying is that we should be fine with join() since the process would be dead already.

good point. I removed the timeout on join.

**Pull request merged**